keps/sig-node/20190626-node-shutdown-state.md

```diff
@@ -0,0 +1,237 @@
+
+The impact of this behaviour wasn’t visible as when a node is shutdown some of the cloud providers (if not most) were deleting the node object. This led the podgc controller to remove orphaned pods using a 0 grace-period[2]. As a consequence, the attach_detach controller received a watch notification for pod deletion and removed that pod and its attachment from its desired state. Finally when reconciling the actual state with desired state, the reconciler triggers the detach.
+
+This behaviour was very racy, as while the node was being deleted, it could come back and register itself, list its pods and start the containers while volumes are being detached, leading to data corruption. To mitigate this, we explicitly set requirements on handling the node objects for cloud providers[3], i.e. not deleting node object unless the instance do not exist anymore.
+
+This made it noticeable to users, several were complaining about it, as a temporary mitigation we introduced taint called node-shutdown enabling cluster administrators to write automation against it, although most of the solutions at the time were racy.
+
+Solving this issue requires also taking into account the existing feature set, the most notable one is the alpha bootstrap checkpointing[4], but since it’s alpha it is not enabled by default which makes the kubelet use usual flow (i.e check in to get the latest state from the apiserver). Also, sig-cluster-lifecycle (which is its main consumer) decided to not pursue this path and start its deprecation[5]
+
+[1] [Pod safety design](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/pod-safety.md)
+[2] [Gc orphaned pods](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/gc_controller.go#L97)
+[3] [requirements on node objects](https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L167)
+[8] [bootstrap checkpointing](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/0004-bootstrap-checkpointing.md)
+[9] [bootstrap checkpointing deprecation intent](https://github.com/kubernetes/enhancements/issues/1103)
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Introduce a new API to report node state
+* Handle cases where the Kubelet itself is unresponsive
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is shutdown without any intervention
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP do not intend to introduce any new API, instead we plan on re-using the Lease API and a mechanism which is similar to leader election (if not the same).
+As stated this KEP is all about finding the right inter-lock between all the actors. The main are:
+
+* Kube-Controller-Manager
+* Kubelets
+
```

Comment:
I would add "user workloads" here.  A workload has to describe the tradeoffs between availability and partition tolerance that it will accept.  That includes:

1. access to exclusive devices/volumes (for instance iSCSI is really an RWM server but they may want kube to enforce RWO)
2. toleration of node shutdown at the pod level (the amount of outage the pod tolerates)
3. the type of workload controller (replica set or stateful set or DaemonSet)
4. additional future signals

The outcome we need to reach to move Kube forward is to

1. Consider a change to the default behavior that tightens the guarantees on pod behavior so pods are still safe by default (no unanticipated behavior) but users don't have to know to ask for better guarantees
2. Provide the tunable that allows a workload to relax / tighten those constraints

Reply:
on 2. how do you see users describing constrains around their workload's guarantees ?

Reply:
It has to be something similar to the existing toleration (unready state) which is applied on the control plane side.  Questions include: do we need a new toleration to be backwards compatible?  is it an error or a feature to have different toleration lengths (tolerate api server down for 600s but tolerate node unready for 30s)?  should stateful sets encourage setting both tolerations?  does setting the toleration change statefulset behavior?

Reply:
for data consistency - SCSI persistent reservations or lockfiles could already be used to provide RWO on a RWM media

Reply:
if the underlying storage doesn't support it, then maybe a mutex could be implemented by the cloud provider / PV plugin.